Stock-Analysis/src/Ksfraser/Finance/2000/OtherSoftware/tsinvest/tslogistic/tslogistic.1
.TH TSLOGISTIC 1 "January 17, 2006"
.SH NAME
tslogistic \- logistic function generator-generates a time series
.SH SYNOPSIS
.nf
tslogistic [-v] a b c number
.fi
.SH DESCRIPTION
Tslogistic is for generating a logistic function time series. The idea
is to produce a function of the form y(t) = c / (1 + e^(-(at + b))).
See "The Art of Modeling Dynamic Systems", Foster Morrison, John Wiley
& Sons, New York, New York, 1991, pp 100, or "Predictions", Theodore
Modis, Simon & Schuster, New York, New York, 1992, ISBN 0-471-52004-7,
pp 230.

Note: these programs use the following functions from other
references:

    ran1, which returns a uniform random deviate between
    0.0 and 1.0. See "Numerical Recipes in C: The Art of
    Scientific Computing," William H. Press, Brian
    P. Flannery, Saul A. Teukolsky, William T. Vetterling,
    Cambridge University Press, New York, 1988, ISBN
    0-521-35465-X, page 210, referencing Knuth.

    gasdev, which returns a normally distributed deviate
    with zero mean and unit variance, using ran1 () as the
    source of uniform deviates. See "Numerical Recipes in
    C: The Art of Scientific Computing," William H. Press,
    Brian P. Flannery, Saul A. Teukolsky, William
    T. Vetterling, Cambridge University Press, New York,
    1988, ISBN 0-521-35465-X, page 217.

    gammln, which returns the log of the results of the
    gamma function.  See "Numerical Recipes in C: The Art
    of Scientific Computing," William H. Press, Brian
    P. Flannery, Saul A. Teukolsky, William T. Vetterling,
    Cambridge University Press, New York, 1988, ISBN
    0-521-35465-X, page 168.
.SH OPTIONS
.TP
.B -t
Sample's time will be included in the output time series.
.TP
.B -v
Print the version and copyright banner of the program.
.TP
.B a, b, c
Logistic function constants.
.TP
.B number
Number of data points in the output time series.
.SH WARNINGS
.PP
There is little or no provision for handling numerical exceptions.
.SH SEE ALSO
.PP
tsderivative(1), tshcalc(1), tshurst(1), tsintegrate(1),
tslogreturns(1), tslsq(1), tsnormal(1), tsshannon(1),
tsblack(1), tsbrownian(1), tsdlogistic(1), tsfBm(1),
tsfractional(1), tsgaussian(1), tsintegers(1),
tslogistic(1), tspink(1), tsunfairfractional(1),
tswhite(1), tscoin(1), tsunfairbrownian(1), tsfraction(1),
tsshannonmax(1), tschangewager(1), tssample(1), tsrms(1),
tscoins(1), tsavg(1), tsXsquared(1), tsstockwager(1),
tsshannonwindow(1), tsmath(1), tsavgwindow(1), tspole(1),
tsdft(1), tsbinomial(1), tsdeterministic(1), tsnumber(1),
tsrmswindow(1), tsshannonstock(1), tsmarket(1),
tsstock(1), tsstatest(1), tsunfraction(1),
tsshannonaggregate(1), tsinstant(1), tsshannonvolume(1),
tsstocks(1), tsshannonfundamental(1), tstrade(1),
tstradesim(1), tsrunlength(1), tsunshannon(1),
tsrootmean(1), tsrunmagnitude(1), tskurtosis(1),
tskurtosiswindow(1), tsrootmeanscale(1),
tsscalederivative(1), tsgain(1), tsgainwindow(1)
tscauchy(1), tslognormal(1), tskalman(1), tsroot(1),
tslaplacian(1)
.SH DIAGNOSTICS
.PP
Error messages for incompatible arguments, failure to allocate memory,
inaccessible files, and opening and closing files.
.SH AUTHORS
.nf
----------------------------------------------------------------------

A license is hereby granted to reproduce this software source code and
to create executable versions from this source code for personal,
non-commercial use.  The copyright notice included with the software
must be maintained in all copies produced.

THIS PROGRAM IS PROVIDED "AS IS". THE AUTHOR PROVIDES NO WARRANTIES
WHATSOEVER, EXPRESSED OR IMPLIED, INCLUDING WARRANTIES OF
MERCHANTABILITY, TITLE, OR FITNESS FOR ANY PARTICULAR PURPOSE.  THE
AUTHOR DOES NOT WARRANT THAT USE OF THIS PROGRAM DOES NOT INFRINGE THE
INTELLECTUAL PROPERTY RIGHTS OF ANY THIRD PARTY IN ANY COUNTRY.

Copyright (c) 1994-2006, John Conover, All Rights Reserved.

Comments and/or bug reports should be addressed to:

    john@email.johncon.com (John Conover)

----------------------------------------------------------------------
.fi


Stock-Analysis/src/Ksfraser/Finance/2000/OtherSoftware/tsinvest/tslogistic/tslogistic.c
/*

------------------------------------------------------------------------------

A license is hereby granted to reproduce this software source code and
to create executable versions from this source code for personal,
non-commercial use.  The copyright notice included with the software
must be maintained in all copies produced.

THIS PROGRAM IS PROVIDED "AS IS". THE AUTHOR PROVIDES NO WARRANTIES
WHATSOEVER, EXPRESSED OR IMPLIED, INCLUDING WARRANTIES OF
MERCHANTABILITY, TITLE, OR FITNESS FOR ANY PARTICULAR PURPOSE.  THE
AUTHOR DOES NOT WARRANT THAT USE OF THIS PROGRAM DOES NOT INFRINGE THE
INTELLECTUAL PROPERTY RIGHTS OF ANY THIRD PARTY IN ANY COUNTRY.

Copyright (c) 1994-2006, John Conover, All Rights Reserved.

Comments and/or bug reports should be addressed to:

    john@email.johncon.com (John Conover)

------------------------------------------------------------------------------

tslogistic.c, logistic function generator-generates a time series.
The idea is to produce a function of the form y(t) = c / (1 + e^(-(at
+ b))).  See "The Art of Modeling Dynamic Systems", Foster Morrison,
John Wiley & Sons, New York, New York, 1991, pp 100, or "Predictions",
Theodore Modis, Simon & Schuster, New York, New York, 1992, ISBN
0-471-52004-7, pp 230.

Note: these programs use the following functions from other
references:

    ran1, which returns a uniform random deviate between 0.0 and
    1.0. See "Numerical Recipes in C: The Art of Scientific
    Computing," William H. Press, Brian P. Flannery, Saul
    A. Teukolsky, William T. Vetterling, Cambridge University Press,
    New York, 1988, ISBN 0-521-35465-X, page 210, referencing Knuth.

    gasdev, which returns a normally distributed deviate with zero
    mean and unit variance, using ran1 () as the source of uniform
    deviates. See "Numerical Recipes in C: The Art of Scientific
    Computing," William H. Press, Brian P. Flannery, Saul
    A. Teukolsky, William T. Vetterling, Cambridge University Press,
    New York, 1988, ISBN 0-521-35465-X, page 217.

    gammln, which returns the log of the results of the gamma
    function.  See "Numerical Recipes in C: The Art of Scientific
    Computing," William H. Press, Brian P. Flannery, Saul
    A. Teukolsky, William T. Vetterling, Cambridge University Press,
    New York, 1988, ISBN 0-521-35465-X, page 168.

$Revision: 0.0 $
$Date: 2006/01/18 20:28:55 $
$Id: tslogistic.c,v 0.0 2006/01/18 20:28:55 john Exp $
$Log: tslogistic.c,v $
Revision 0.0  2006/01/18 20:28:55  john
Initial version


*/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <unistd.h>

static char rcsid[] = "$Id: tslogistic.c,v 0.0 2006/01/18 20:28:55 john Exp $"; /* program version */
static char copyright[] = "Copyright (c) 1994-2006, John Conover, All Rights Reserved"; /* the copyright banner */

#ifdef __STDC__

static const char *help_message[] = /* help message index array */

#else

static char *help_message[] = /* help message index array */

#endif

{
    "\n",
    "Generate a logistic function time series\n",
    "Usage: tslogistic [-t] [-v] a b c number\n",
    "where the logistic function is y(t) = c / (1 + e^(-(at + b)))\n",
    "    -t, sample's time will be included in the output time series\n",
    "    -v, print the program's version information\n",
    "    number is the number of samples in the time series\n"
};

#ifdef __STDC__

static const char *error_message[] = /* error message index array */

#else

static char *error_message[] = /* error message index array */

#endif

{
    "No error\n",
    "Error in program argument(s)\n"
};

#define NOERROR 0 /* error values, one for each index in the error message array */
#define EARGS 1

#ifdef __STDC__

static void print_message (int retval); /* print any error messages */

#else

static void print_message (); /* print any error messages */

#endif

#ifdef __STDC__

int main (int argc,char *argv[])

#else

int main (argc,argv)
int argc;
char *argv[];

#endif

{
    int number, /* number of records in time series */
        retval = EARGS, /* return value, assume not enough arguments */
        i, /* counter of number of records in time series */
        t = 0, /* print time of samples flag, 0 = no, 1 = yes */
        c; /* command line switch */

    double a, /* first parameter of logistic equation */
           b, /* second parameter of logistic equation */
           s; /* scaling factor of logistic equation */

    while ((c = getopt (argc, argv, "s:tv")) != EOF) /* for each command line switch */
    {

        switch (c) /* which switch? */
        {

            case 't': /* request printing time of samples? */

                t = 1; /* yes, set the print time of samples flag */
                break;

            case 'v':

                (void) printf ("%s\n", rcsid); /* print the version */
                (void) printf ("%s\n", copyright); /* print the copyright */
                optind = argc; /* force argument error */

            default: /* illegal switch? */

                optind = argc; /* force argument error */
                break;
        }

    }

    if (argc - optind > 3) /* enough arguments? */
    {
        retval = NOERROR; /* assume no error */

        a = atof (argv[optind]); /* first parameter of logistic equation */
        b = atof (argv[optind + 1]); /* second parameter of logistic equation */
        s = atof (argv[optind + 2]); /* scaling factor of logistic equation */
        number = atoi (argv[optind + 3]); /* number of records in time series */

        for (i = 0; i < number; i ++) /* for each record in the time series */
        {

            if (t == 1) /* print time of samples? */
            {
                (void) printf ("%d\t", i); /* yes, print the sample's time */
            }

            (void) printf ("%f\n", s / ((double) 1.0 + exp (-(b + (a * (double) i))))); /* print the record */
        }

    }

    print_message (retval); /* print any error messages */
    exit (retval); /* exit with the error value */

#ifdef LINT

    return (0); /* for lint formalities */

#endif

}

/*

Print any error messages.

static void print_message (int retval);

I) Data structures:

    A) The help_message array is an array of character stings, one per
       line to be printed for help.

    B) The error_message array is an array of character strings, one
       line per error; the error_message array is implicitly addressed
       by the value integer retval, which specifies the error to be
       printed.

II) Function execution:

    A) Depending on the value of the integer retval:

        1) If retval is zero, print nothing-normal/successful program
           exit.

        2) Else if retval is unity, print help.

        3) Else retval is not zero or unity, it is an error code, print
           the corresponding error message.

Returns nothing.

*/

#ifdef __STDC__

static void print_message (int retval)

#else

static void print_message (retval)
int retval;

#endif

{
    size_t help_ctr; /* help_message line counter */

    switch (retval) /* which return value */
    {

        case 0: /* program ended without errors, print nothing */

            break;

        case 1: /* program ended with a request for help, or argument error, print help */

            for (help_ctr = 0; help_ctr < (sizeof (help_message) / sizeof (char *)); help_ctr ++) /* for each line of help */
            {
                (void) printf ("%s", help_message[help_ctr]); /* print the line */
            }

            break;

        default: /* an error that was not a request for help, print the error */

            (void) fprintf (stderr, "%s", error_message[retval]);
            break;

    }

}
